gchq/CyberChef
# Implement Node API

Core wrapping of operations into a callable format has initial implementation.

- Core API functionality exists
- Added `help` function which returns info on operation. Currently accepts string, returns one operation's metadata
- Test capability added

To do:
- [x] Add `Recipe` or `chef.bake` implementation that takes recipe from web app JSON output
- [x] Help - try and use OperationConfig.json. It could be loaded in statically to avoid import issues we are having
- [x] return result as Dish - overwrite toString, toJSON and anything elser that console.log, builtin data types work.
- [x] Remove branching, etc. functions that will not be relevant
- [x] Export individual operations alongside chef (how to do this without manually declaring each op in the top level export?) e.g. `import { fromBase32 } from 'cyberChef'`
- [x] Test type conversion that uses Dish (test `chef.translateTo`)
- [x] Ensure errors come out of operations correctly (not just strings)
- [x] Consider `result` object, which has type conversion - `chef.translateTo` added
- [x] Dont decapitalise all caps functions e.g. ROT13
- [x] Use Objects for optional args rather than arrays
- [x] Add node API to docs
- [x] Add ability to use with callbacks as well as promises

Bugs
- [x] toBase64 shows no output?
- [x] crypto-api error when running build from node
- [x] ES5: `help` and `bake` are not part of the `chef` object - only in `.default`.

To check before release:
- [x] Node support (try in later versions than current)
- [x] One off test of each operation
- [x] Add link to wiki to docs and update wiki
- [x] What do we want in our release? src and build? es6 and es5 builds?
- [x] Diff , Regex and any other visual ops - how to summarise result?
- [x] Consider changing toggleString arg properties? `string` could be `value`
- [x] Generate REPL index - index could be reduced?

Feedback:
- [x] `help` could use search in the same way as the UI. Check for logic. e.g. `chef.help("base64")` and get a sensible answer.
- [x] ensure that `bake` is forgiving: accept "to base 64", "tobase64", etc. Use default args if none specified.

Future issues:
- Add operation as input to `help`, so you can do `chef.help(chef.toBase32)` 
- Add `search` functionality
- Add `run` to `help` output
